### ISSUE TITLE
Very long unit animations cause the game to crash

### ISSUE THREAD

**Issue by inferno8** [wesnoth/wesnoth#2915]
Copied from:
https://forums.wesnoth.org/viewtopic.php?f=4&t=48005

This is probably very UMC-specific, but is related to the way the game handles animation so I decided to report it anyway.

Playing very long unit animations (400+ frames) leads to the following error:
`Could not create a new surface in get_surface_portion()`

The game crashes right after that or it is still running but cannot read the rest of images an animation is composed of, throwing `Surface lost` error to the console.

As far as I know, the error was observed on machines running BfW version: **1.13.14** SDL: 2.0.4 and BfW **1.12.6**, Windows 10 (10.0.16299).

I am fully aware almost no one would use huge animations like these, so this may not be relevant to the core Wesnoth. Nevertheless to help reproduce the problem I created a small test add-on in from of a multiplayer scenario for those who are interested:
[Long_Animation_Test.zip](https://github.com/wesnoth/wesnoth/files/1927082/Long_Animation_Test.zip)

**Comment by Vultraz** [comment]
I... how... would you ever need such a long animation.... O_O

**Comment by Vultraz** [comment]
A very quick look into this seems to indicate the problem is you're hitting Wesnoth's memory limit, as Wesnoth is still built as a 32 bit app on Windows. We're working on remedying that...

**Comment by shikadiqueen** [comment]
You can get this problem on Linux too with higher numbers if you insist on building Wesnoth as an IA32 binary.

Also, it's not specific to the animation engine, it just happens to be a facilitator. The problem is that Wesnoth will try to cache every image it ever needs to load, and the images in question are *huge*, with each one taking up over 1.8 megabytes of memory without taking into account additional overhead from the attributes SDL needs to handle in memory, as well as Wesnoth's image cache structure. Have 1,000 of them and you start to run out of RAM — 32-bit Windows processes have a 2 GB limit for user mode mappings by default, and some of that memory is needed for libraries and memory-mapped files, as well as the execution stack and some miscellaneous structures.

**Comment by inferno8** [comment]
> I... how... would you ever need such a long animation.... O_O

Such long animations are used in one of my add-ons to imitate cutscenes. The game does not support videos (yet) so this is a workaround. I use a dummy unit with [extra_anim] for this.

**Comment by CelticMinstrel** [comment]
This may just go away with OpenGL, though if you're using screen-size images here to simulate a video, it could still be an issue…

**Comment by hrubymar10** [comment]
And wesnoth can't clean it's memory after exceeding for example 1.5Gb?

After testing attached Long_Animation_Test:
<img width="1063" alt="image" src="https://user-images.githubusercontent.com/2218817/39084975-304337b2-457d-11e8-8499-ccb86138c281.png">

**Comment by AI0867** [comment]
This issue is essentially why video encodings are a lot more complicated than "a series of images".

As for whether wesnoth can clean its memory, most of the images it loads are *tiny*, and disk access is far more expensive than just keeping it in RAM forever (though I'm not sure if that's what we actually do for everything). There are some exceptions (story screens), but you've pretty much written a pathological case.

**Comment by jostephd** [comment]
So.. what is this issue asking for? A way for scenario authors to mark some images as "don't cache these"? A way to play videos during scenarios? Having get_surface_portion's callers do nullptr checks before using the returned value?

**Comment by ProditorMagnus** [comment]
Video is #2918

**Comment by Vultraz** [comment]
Mitigated by the switch to 64 bit, at least.

**Comment by Pentarctagon** [comment]
So is this essentially resolved with 1.16 since all official executables are now 64-bit, or is there something else that needs to change (such as clearing out older/bigger cached images at a certain point)?

**Comment by irydacea** [comment]
> So is this essentially resolved with 1.16 since all official executables are now 64-bit, or is there something else that needs to change (such as clearing out older/bigger cached images at a certain point)?

I'd say yes to the latter, unless you like your games to consume more memory than you have RAM+swap space available.

**Comment by Wedge009** [comment]
If this is memory-usage related, my guess is #6030 could be related as well.

**Comment by Pentarctagon** [comment]
The path taken to make Wesnoth use a ton of RAM is different there, but the cause very likely the same, yeah.

**Comment by Pentarctagon** [comment]
So avoiding caching large images is pretty easy, but for some reason not caching the animation image frames causes fps to drop from 15-17 to 10-11 (debug build).  Given the animation frames are only used once, I wouldn't think caching them would've made a difference, since they would never have been used a second time to be retrieved from the cache anyway.

**Comment by cooljeanius** [comment]
#6030 is about terrain animations, not unit animations, so maybe the title should be updated?

**Comment by AI0867** [comment]
Regarding an approach to this, it would be nice to have something like android's `onTrimMemory()`, but I suppose we'll have to write our own heuristics or hard-coded limits.

An LRU approach would probably be fine, possibly combined with weighting towards larger images.

**Comment by jimis** [comment]
> An LRU approach would probably be fine, possibly combined with weighting towards larger images.

It is possible to take advantage of the OS's LRU cache, if wesnoth's image cache directly points to raw image files from disk mmap()ed read-only. Under memory pressure the OS will unload the least recently used files.

P.S. I ended up in this bug report after seeing Wesnoth 1.16.1 taking more than 2GB resident memory on the simplest scenarios, competing with other processes on my old system with 4GB RAM. Not sure if the "image cache" is at fault though, not without first profiling memory usage.

**Comment by Pentarctagon** [comment]
To clarify, the cache is of SDL surfaces, not of the actual JPG/PNG images.

**Comment by jimis** [comment]
I took a memory profile of Wesnoth 1.6.1 after loading the last turn of Tutorial Part2 scenario, where I noticed the RSS is 1.2GB.

More than half of the memory has been allocated from within get_scaled_to_hex() inside image::get_image(). The rest mostly comes from get_hexed() in the same function. I also manually verified that zooming-in indeed increases memory consumption a lot, and I tend to play zoomed in 3x.

I attach a screenshot of the flamegraph. I can post more details here or in a new issue if there is interest.

![wesnoth_tutorial_part2_zoomed3x heaptrack-flamegraph screenshot](https://user-images.githubusercontent.com/3584514/145213161-3666f25a-53e0-4cc4-872f-07c6044ae807.png)

**Comment by jimis** [comment]
OK, just let me know if you want to see other parts of the memory profile. I think this flamegraph says it all, but some parts of it are not clearly visible. It's much more clear to me that I can interact with the tool.

BTW I understand that scaling in Wesnoth happens in software. It is fractional (floating point) or integral scaling? Is it so CPU intensive that it can't happen on demand?

**Comment by Pentarctagon** [comment]
I don't know.

**Comment by Pentarctagon** [comment]
As an update to this - the caching of scaled images has been removed, so that part should no longer be an issue.

**Comment by cooljeanius** [comment]
> As an update to this - the caching of scaled images has been removed, so that part should no longer be an issue.

...that was in @mesilliac's graphics overhaul, correct?

**Comment by Pentarctagon** [comment]
Right.

**Comment by jimis** [comment]
Memory usage is lower now right after loading a scenario, and seems unaffected by the zoom level. In particular, loading the same savegame I mentioned above (the last turn of Tutorial Part2), and scrolling around, RSS was at ~550MB which is a great improvement over the previous 1.2 GB.

OTOH it seems memory is never freed, not even when I exit to the main menu, so loading and playing different savegames caused the memory usage to grow to more than 1GB. This might be a different issue though.

**Comment by Vultraz** [comment]
FTR, the cache mechanism was improved extensively in 9699377ee08996380daad146e641b60a659ec719 (ironically only 3 days after the above comment). Still need to add a max-size mechanism.

**Comment by mentos987** [comment]
Would converting such big animations to .gif help?

**Comment by AI0867** [comment]
I doubt it. The engine caches *SDL surfaces*, not the image files themselves, so the format the images are stored in isn't actually that relevant.

**Comment by gfgtdf** [comment]
coming up with a good logic on how exactly to clear the cache might be quite hard to do, what woud be rather easy do to however is to add way to umc to signal the engine to disable caching for a certain image, for example via a custom image path function ~NOCACHE() that wouldn't effect how thefinal image looks, just effect how it's cached

**Comment by Pentarctagon** [comment]
Probably something that's a combination of size (don't keep huge surfaces in the cache for too long), last access time (remove things that haven't been accessed recently), and whitelisting either certain sources from being removed (ie: terrain) or surfaces under a certain size would work pretty well.

**Comment by mentos987** [comment]
If we remove cache it will cause these large images to be reloaded from scratch on every frame update. Will that not be rather heavy? would something like .gif help in that regard?

**Comment by Pentarctagon** [comment]
As AI0867 said, the image format is irrelevant.

**Comment by mentos987** [comment]
> The engine caches SDL surfaces

Yes, I get that. But let us assume that we don't cache any of the bigger animations. That would mean that we have to redraw/rebuild them from scratch on every frame update. Are not gifs more efficient at that?

**Comment by Pentarctagon** [comment]
A gif image can't be converted to a surface. Each frame in the gif would be its own surface, exactly the same as it is currently.
